Bug: A C# comment with HORIZONTAL ELLIPSIS (U+2026) results in incorrect parse

Compile this:

    // Test.cs
    // Hello … world

Result:

    csc.exe /target:library Test.cs
    Microsoft (R) Visual C# Compiler version 1.0.0.50415
    Copyright (C) Microsoft Corporation. All rights reserved.

    test.cs(3,2): error CS0116: A namespace cannot directly contain members such as fields or methods

Expect:
Compiles with no error.

--- a/src/Compilers/Core/CodeAnalysisTest/Text/EncodedStringTextTests.cs
+++ b/src/Compilers/Core/CodeAnalysisTest/Text/EncodedStringTextTests.cs
@@ -97,7 +97,7 @@ namespace Microsoft.CodeAnalysis.UnitTests
         public void Decode_NonUtf8()
         {
             var utf8 = new UTF8Encoding(false, true);
-            var encoding = Encoding.GetEncoding(name: "Latin1");
+            var encoding = Encoding.Default;
             var text = "abc def baz aeiouy " + encoding.GetString(new byte[] { 0x80, 0x92, 0xA4, 0xB6, 0xC9, 0xDB, 0xED, 0xFF });
             var bytes = encoding.GetBytesWithPreamble(text);
 
@@ -112,12 +112,12 @@ namespace Microsoft.CodeAnalysis.UnitTests
                 Assert.True(stream.CanRead);
             }
 
-            // Detect encoding should correctly pick Encoding.Default
+            // Detect encoding should correctly pick Encoding.Default (CodePage 1252)
             using (var stream = new MemoryStream(bytes))
             {
                 var sourceText = EncodedStringText.Create(stream);
                 Assert.Equal(text, sourceText.ToString());
-                Assert.Equal(encoding, sourceText.Encoding);
+                Assert.Equal(encoding.CodePage, sourceText.Encoding.CodePage);
                 Assert.True(stream.CanRead);
             }
         }

--- a/src/Compilers/Core/CodeAnalysisTest/Text/EncodedStringTextTests.cs
+++ b/src/Compilers/Core/CodeAnalysisTest/Text/EncodedStringTextTests.cs
@@ -96,12 +96,19 @@ namespace Microsoft.CodeAnalysis.UnitTests
         [Fact]
         public void Decode_NonUtf8()
         {
-            var utf8 = new UTF8Encoding(false, true);
-            var encoding = Encoding.Default;
-            var text = "abc def baz aeiouy " + encoding.GetString(new byte[] { 0x80, 0x92, 0xA4, 0xB6, 0xC9, 0xDB, 0xED, 0xFF });
-            var bytes = encoding.GetBytesWithPreamble(text);
+            // Unicode text with extended characters that map to interesting code points in CodePage 1252.
+            var text = "abc def baz aeiouy \u20ac\u2019\u00a4\u00b6\u00c9\u00db\u00ed\u00ff";
 
-            // Encoding.Default should not decode to UTF-8
+            // The same text encoded in CodePage 1252 which happens to be an illegal sequence if decoded as Utf-8.
+            var bytes = new byte[]
+            {
+                0x61, 0x62, 0x63, 0x20, 0x64, 0x65, 0x66, 0x20, 0x62, 0x61, 0x7a, 0x20, 0x61, 0x65, 0x69, 0x6f, 0x75, 0x79, 0x20,
+                0x80, 0x92, 0xA4, 0xB6, 0xC9, 0xDB, 0xED, 0xFF
+            };
+
+            var utf8 = new UTF8Encoding(false, true);
+
+            // bytes should not decode to UTF-8
             using (var stream = new MemoryStream(bytes))
             {
                 Assert.Throws(typeof(DecoderFallbackException), () =>
@@ -112,12 +119,12 @@ namespace Microsoft.CodeAnalysis.UnitTests
                 Assert.True(stream.CanRead);
             }
 
-            // Detect encoding should correctly pick Encoding.Default (CodePage 1252)
+            // Detect encoding should correctly pick CodePage 1252
             using (var stream = new MemoryStream(bytes))
             {
                 var sourceText = EncodedStringText.Create(stream);
                 Assert.Equal(text, sourceText.ToString());
-                Assert.Equal(encoding.CodePage, sourceText.Encoding.CodePage);
+                Assert.Equal(1252, sourceText.Encoding.CodePage);
                 Assert.True(stream.CanRead);
             }
         }

--- a/src/Compilers/Core/CodeAnalysisTest/Text/EncodedStringTextTests.cs
+++ b/src/Compilers/Core/CodeAnalysisTest/Text/EncodedStringTextTests.cs
@@ -266,5 +266,19 @@ namespace Microsoft.CodeAnalysis.UnitTests
                 Assert.Equal(0, encodedText.Length);
             }
         }
+
+        [Fact, WorkItem(2081, "https://github.com/dotnet/roslyn/issues/2081")]
+        public void HorizontalEllipsis()
+        {
+            // Character 0x85 in CodePage 1252 is a horizontal ellipsis.
+            // If decoded as Latin-1, then it's incorrectly treated as \u0085 which
+            // is a line break ('NEXT LINE').
+            byte[] srcBytes = new[] { (byte)0x85 };
+            using (var ms = new MemoryStream(srcBytes))
+            {
+                var sourceText = EncodedStringText.Create(ms);
+                Assert.Equal('\u2026', sourceText[0]);
+            }
+        }
     }
 }

--- a/src/Compilers/Core/Desktop/CodeAnalysis.Desktop.csproj
+++ b/src/Compilers/Core/Desktop/CodeAnalysis.Desktop.csproj
@@ -57,6 +57,7 @@
       <DesignTime>True</DesignTime>
       <DependentUpon>CodeAnalysisDesktopResources.resx</DependentUpon>
     </Compile>
+    <Compile Include="CodePage1252Encoding.cs" />
     <Compile Include="CommandLine\CommandLineReference.cs" />
     <Compile Include="CommandLine\CommandLineSourceFile.cs" />
     <Compile Include="CommandLine\CommonCommandLineArguments.cs" />

--- a/src/Compilers/Core/Desktop/CodePage1252Encoding.cs
+++ b/src/Compilers/Core/Desktop/CodePage1252Encoding.cs
@@ -1,0 +1,123 @@
+﻿using System;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Text
+{
+    /// <summary>
+    /// CodePage 1252 decoder. This is a single byte character set encoding that is very close
+    /// to Latin1 except some of the code points in the control set 1 range (0x80 to 0x9F) are
+    /// replaced with typographic characters.
+    /// </summary>
+    internal class CodePage1252Encoding : Encoding
+    {
+        public static readonly CodePage1252Encoding Instance = new CodePage1252Encoding();
+
+        public override int CodePage => 1252;
+
+        public override int GetByteCount(char[] chars, int index, int count)
+        {
+            // This is a decoder only
+            throw new NotSupportedException();
+        }
+
+        public override int GetBytes(char[] chars, int charIndex, int charCount, byte[] bytes, int byteIndex)
+        {
+            // This is a decoder only
+            throw new NotSupportedException();
+        }
+
+        // This is SBCS, so the number of characters is the same as the number of bytes.
+        public override int GetCharCount(byte[] bytes, int index, int count) => count;
+
+        // This is SBCS, so the number of characters is the same as the number of bytes.
+        public override unsafe int GetCharCount(byte* bytes, int count) => count;
+
+        public override int GetChars(byte[] bytes, int byteIndex, int byteCount, char[] chars, int charIndex)
+        {
+            int byteEndIndex = byteIndex + byteCount;
+            while (byteIndex < byteEndIndex)
+            {
+                chars[charIndex++] = Decode(bytes[byteIndex++]);
+            }
+
+            return byteCount;
+        }
+
+        public override unsafe int GetChars(byte* bytes, int byteCount, char* chars, int charCount)
+        {
+            if (charCount < byteCount)
+            {
+                throw new ArgumentException(nameof(charCount)); // The output char buffer is too small to contain the decoded characters
+            }
+
+            byte* end = bytes + byteCount;
+            while (bytes < end)
+            {
+                *chars++ = Decode(*bytes++);
+            }
+
+            return byteCount;
+        }
+
+        private static readonly char[] s_c1ControlChars =
+        {
+            /* x80 */ '\u20ac' /* Euro Sign */,
+            /* x81 */ '\u0081' /* <control> */,
+            /* x82 */ '\u201a' /* Colon Sign */,
+            /* x83 */ '\u0192' /* Latin Small Letter F With Hook */,
+            /* x84 */ '\u201e' /* Double Low-9 Quotation Mark */,
+            /* x85 */ '\u2026' /* Horizontal Ellipsis */,
+            /* x86 */ '\u2020' /* Dagger */,
+            /* x87 */ '\u2021' /* Double Dagger */,
+            /* x88 */ '\u02c6' /* Modifier Letter Cicumflex Accent */,
+            /* x89 */ '\u2030' /* Per Mille Sign */,
+            /* x8a */ '\u0160' /* Latin Capital Letter S With Caron */,
+            /* x8b */ '\u2039' /* Single Left-Pointing Angle Quotation Mark */,
+            /* x8c */ '\u0152' /* Latin Capital Ligature Oe */,
+            /* x8d */ '\u008d' /* Reverse Line Feed */,
+            /* x8e */ '\u017d' /* Latin Capital Letter Z With Caron */,
+            /* x8f */ '\u008f' /* Single Shift Three */,
+            /* x90 */ '\u0090' /* Device Control String */,
+            /* x91 */ '\u2018' /* Left Single Quotation Mark */,
+            /* x92 */ '\u2019' /* Right Single Quotation Mark */,
+            /* x93 */ '\u201c' /* Left Double Quotation Mark */,
+            /* x94 */ '\u201d' /* Right Double Quotation Mark */,
+            /* x95 */ '\u2022' /* Bullet */,
+            /* x96 */ '\u2013' /* En Dash */,
+            /* x97 */ '\u2014' /* Em Dash */,
+            /* x98 */ '\u02dc' /* Small Tilde */,
+            /* x99 */ '\u2122' /* Trade Mark Sign */,
+            /* x9a */ '\u0161' /* Latin Small Letter S With Caron */,
+            /* x9b */ '\u203a' /* Single Right-Pointing Angle Quotation Mark */,
+            /* x9c */ '\u0153' /* Latin Small Ligature Oe */,
+            /* x9d */ '\u009d' /* Operating System Command */,
+            /* x9e */ '\u017e' /* Latin Small Letter Z With Caron */,
+            /* x9f */ '\u0178' /* Latin Captial Letter Y With Diaeresis */,
+        };
+
+        private static char Decode(byte b)
+        {
+            // For characters in the C1 Control set (0x80 to 0x9F), use the table.
+            // Otherwise, the mapping is 1 to 1.
+            uint c1 = unchecked(b - 0x80u);
+            return c1 < 0x20u ? s_c1ControlChars[c1] : (char)b;
+        }
+
+        public override int GetMaxByteCount(int charCount)
+        {
+            // This is a decoder only
+            throw new NotSupportedException();
+        }
+
+        public override int GetMaxCharCount(int byteCount)
+        {
+            if (byteCount < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(byteCount));
+            }
+
+            // This is SBCS, so everything is 1:1
+            return byteCount;
+        }
+    }
+}

--- a/src/Compilers/Core/Desktop/CodePage1252Encoding.cs
+++ b/src/Compilers/Core/Desktop/CodePage1252Encoding.cs
@@ -8,7 +8,7 @@ namespace Microsoft.CodeAnalysis.Text
     /// to Latin1 except some of the code points in the control set 1 range (0x80 to 0x9F) are
     /// replaced with typographic characters.
     /// </summary>
-    internal class CodePage1252Encoding : Encoding
+    internal sealed class CodePage1252Encoding : Encoding
     {
         public static readonly CodePage1252Encoding Instance = new CodePage1252Encoding();
 

--- a/src/Compilers/Core/Desktop/EncodedStringText.cs
+++ b/src/Compilers/Core/Desktop/EncodedStringText.cs
@@ -23,7 +23,7 @@ namespace Microsoft.CodeAnalysis.Text
         /// from <see cref="SourceText.From(Stream, Encoding, SourceHashAlgorithm, bool)"/> in two ways:
         /// 1. It attempts to minimize allocations by trying to read the stream into a byte array.
         /// 2. If <paramref name="defaultEncoding"/> is null, it will first try UTF8 and, if that fails, it will
-        ///    try extended ASCII (code page Latin1).
+        ///    try CodePage 1252.
         /// </summary>
         /// <param name="stream">The stream containing encoded text.</param>
         /// <param name="defaultEncoding">
@@ -57,7 +57,7 @@ namespace Microsoft.CodeAnalysis.Text
 
             try
             {
-                return Decode(stream, defaultEncoding ?? Encoding.GetEncoding(name: "Latin1"), checksumAlgorithm, throwIfBinaryDetected: detectEncoding);
+                return Decode(stream, defaultEncoding ?? CodePage1252Encoding.Instance, checksumAlgorithm, throwIfBinaryDetected: detectEncoding);
             }
             catch (DecoderFallbackException e)
             {

--- a/src/Workspaces/Core/Desktop/Workspaces.Desktop.csproj
+++ b/src/Workspaces/Core/Desktop/Workspaces.Desktop.csproj
@@ -62,6 +62,9 @@
     <Compile Include="..\..\..\Compilers\Core\Desktop\AssemblyReferenceResolver.cs">
       <Link>InternalUtilities\AssemblyReferenceResolver.cs</Link>
     </Compile>
+    <Compile Include="..\..\..\Compilers\Core\Desktop\CodePage1252Encoding.cs">
+      <Link>InternalUtilities\CodePage1252Encoding.cs</Link>
+    </Compile>
     <Compile Include="..\..\..\Compilers\Core\Desktop\MetadataFileReferenceResolver.cs">
       <Link>InternalUtilities\MetadataFileReferenceResolver.cs</Link>
     </Compile>

